### PR TITLE
[optimize](vectorization)Optimize to_string's performance.

### PR DESCRIPTION
**Pull request by zclllyybb**
## Problem summary

When we call `DataType::to_string(IColumn, row_num)` for `column_const`, the code now would convert const column to full column. This is expensive and necessary. This pr removed it.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [x] Is this PR support rollback (If NO, please explain WHY)

### CODE REVIEW

File: be/src/vec/data_types/data_type_decimal.cpp [hunk 1]
```diff
@@ -45,6 +47,10 @@ bool DataTypeDecimal<T>::equals(const IDataType& rhs) const {
 
 template <typename T>
 std::string DataTypeDecimal<T>::to_string(const IColumn& column, size_t row_num) const {
+    if (typeid(column) == typeid(ColumnConst))
```
Comment (HappenLee) [inline be/src/vec/data_types/data_type_decimal.cpp:50]: Do not use typeid, use `check_and_get_column`

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

File: be/src/vec/data_types/data_type_array.cpp [hunk 2]
```diff
@@ -110,12 +113,14 @@ void get_decimal_value(const IColumn& nested_column, DecimalV2Value& decimal_val
...
-    auto ptr = column.convert_to_full_column_if_const();
-    auto& data_column = assert_cast<const ColumnArray&>(*ptr.get());
+    auto result = check_column_const_set_readability(column, row_num);
+    ColumnPtr ptr = result.first;
+    row_num = result.second;
+
+    auto& data_column = assert_cast<const ColumnArray&>(*ptr);
     auto& offsets = data_column.get_offsets();
 
     size_t offset = offsets[row_num - 1];
```
Comment (HappenLee) [inline be/src/vec/data_types/data_type_array.cpp:126]: row_num is `size_t`  is danger to do `row_num - 1` recheck the logic

File: be/src/vec/data_types/data_type_decimal.cpp [hunk 3]
```diff
@@ -45,25 +51,31 @@ bool DataTypeDecimal<T>::equals(const IDataType& rhs) const {
...
 
 template <typename T>
 void DataTypeDecimal<T>::to_string(const IColumn& column, size_t row_num,
                                    BufferWritable& ostr) const {
-    // TODO: Reduce the copy in std::string mem to ostr, like DataTypeNumber
+    auto result = check_column_const_set_readability(column, row_num);
+    ColumnPtr ptr = result.first;
+    row_num = result.second;
+
+    fmt::memory_buffer buffer;
```
Comment (HappenLee) [inline be/src/vec/data_types/data_type_decimal.cpp:69]: seems take more mem alloc, not better than old

Comment (HappenLee) [inline be/src/vec/data_types/data_type_decimal.cpp:69]: seems take more mem alloc, not better than old

**Comment by zclllyybb** [comment]
run p0

**Comment by zclllyybb** [comment]
run buildall

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**